openhtf/util/console_output.py

```diff
@@ -0,0 +1,253 @@
+"""
+
+import logging
+import math
+import sys
+import time
+
+import colorama
+import contextlib2 as contextlib
+
+# Colorama module has to be initialized before use.
+colorama.init()
+
+_LOG = logging.getLogger(__name__)
+
+
+class ActionFailedError(Exception):
+  """Raised (and caught) when an action is marked as failed."""
+
+
+def banner_print(msg, color='', width=60, file=sys.stdout):
+  """Print the message as a banner with a fixed width.fixed
+
+  Args:
+    msg: The message to print.
+    color: Optional colorama color string to be applied to the message. You can
+        concatenate colorama color strings together in order to get any set of
+        effects you want.
+    width: Total width for the resulting banner.
+    file: A file object to which the banner text will be written. Intended for
+        use with CLI output file objects like sys.stdout.
+
+  Example:
+
+    >>> banner_print('Foo Bar Baz')
+    ======================== Foo Bar Baz =======================
+  """
+  lpad = int(math.ceil((width - len(msg) - 2) / 2.0)) * '='
+  rpad = int(math.floor((width - len(msg) - 2) / 2.0)) * '='
+  file.write('{color}{lpad} {msg} {rpad}\n'.format(
+      color=color, lpad=lpad, msg=msg, rpad=rpad))
+  file.write(colorama.Style.RESET_ALL)
+  file.flush()
+
+
+def bracket_print(msg, color='', width=8, file=sys.stdout):
+  """Prints the message in brackets in the specified color and end the line.
+
+  Args:
+    msg: The message to put inside the brackets (a brief status message).
+    color: Optional colorama color string to be applied to the message. You can
+        concatenate colorama color strings together in order to get any set of
+        effects you want.
+    width: Total desired width of the bracketed message.
+    file: A file object to which the baracketed text will be written. Intended
+        for use with CLI output file objects like sys.stdout.
+    """
+  lpad = int(math.ceil((width - 2 - len(msg)) / 2.0)) * ' '
+  rpad = int(math.floor((width - 2 - len(msg)) / 2.0)) * ' '
+  file.write('[{lpad}{bright}{color}{msg}{reset}{rpad}]'.format(
+      lpad=lpad, bright=colorama.Style.BRIGHT, color=color, msg=msg,
+      reset=colorama.Style.RESET_ALL, rpad=rpad))
+  file.write(colorama.Style.RESET_ALL)
+  file.write('\n')
+  file.flush()
+
+
+def cli_print(msg, color='', file=sys.stdout, logger=_LOG):
+  """Print the message to file and also log it.
+
+  This function is intended as a 'tee' mechanism to enable the CLI interface as
+  a first-class citizen, while ensuring that everything the operator sees also
+  has an analogous logging entry in the test record for later inspection.
+
+  Args:
+    msg: The message to print/log.
+    color: Optional colorama color string to be applied to the message. You can
+        concatenate colorama color strings together in order to get any set of
+        effects you want.
+    file: A file object to which the baracketed text will be written. Intended
```

Comment:
bracketed

(I miss Obama too)

Reply:
Done. Now excuse me while I quietly sob in the corner for a while.

Reply:
Now Joe's crying! Thanks Obama!